gmt-4.5.14/src/mgd77/mgd77convert.1
.TH MGD77CONVERT 1 "1 Nov 2015" "GMT 4.5.14" "Generic Mapping Tools"
.SH NAME
mgd77convert \- Translate between different formats of MGD77 files
.SH SYNOPSIS
\fBmgd77convert\fP \fINGDC-ids\fP \fB\-Fa\fP|\fBc\fP|\fBm\fP|\fBt\fP \fB\-T\fP[\fB+\fP]\fBa\fP|\fBb\fP|\fBm\fP|\fBt\fP 
[ \fB\-L\fP[\fBw\fP][\fBe\fP][\fB+\fP] ] [ \fB\-V\fP ] [ \fB\-4\fP ]
.SH DESCRIPTION
\fBmgd77convert\fP reads versions of MGD77 files and writes the same
data in (probably) another format to a new file in the current directory.  Both pre-
and post-Y2K MGD77 formats can be processed.
.TP
\fINGDC-ids\fP
Can be one or more of five kinds of specifiers:
.br
1) 8-character NGDC IDs, e.g., 01010083, JA010010etc., etc.
.br
2) 2-character <agency> codes which will return all cruises from each agency.
.br
3) 4-character <agency><vessel> codes, which will return all cruises from those vessels.
.br
4) =<list>, where <list> is a table with NGDC IDs, one per line.
.br
5) If nothing is specified we return all cruises in the data base.
.br
(See mgd77info \fB\-L\fP for agency and vessel codes).
The ".mgd77" or ".nc" extensions will automatically be appended, if needed (use \fB\-I\fP
to ignore certain file types).  Cruise files will be looked for first in the current directory
and second in all directories listed in \fB$MGD77_HOME\fP/mgd77_paths.txt [If \fB$MGD77_HOME\fP
is not set it will default to \fB$GMT_SHAREDIR\fP/mgd77].
.br
.TP
\fB\-F\fP
Specifies the format of the input (From) files.  Choose from \fBa\fP for standard
MGD77 ASCII table (with extension .mgd77), \fBc\fP for the new MGD77+ netCDF format
(with extension .nc), revised MGD77T (tab-delimited; with extension .m77t) and \fBt\fP for a plain ASCII tab-separated table dump
(with extension .dat).  Use \fB\-FC\fP to recover the original MGD77 setting from
the MGD77+ file [Default will apply any E77 corrections encoded in the file].
.TP
\fB\-T\fP
Specifies the format of the output (To) files.  Choose from \fBa\fP for standard
MGD77 ASCII table (with extension .mgd77), \fBc\fP for the new MGD77+ netCDF format
(with extension .nc), revised MGD77T (tab-delimited; with extension .m77t) and \fBt\fP for a plain ASCII tab-separated table dump
(with extension .dat).  We will refuse to create the file(s) if they already
exist in the current directory.  Prepend \fB+\fP to override this policy.
.SH OPTIONS
No space between the option flag and the associated arguments.
.TP
\fB\-L\fP
Set the level of verification reporting [none] and where to send such reports [stderr].
Append a combination of \fBw\fP for warnings, \fBe\fP for errors, and \fB+\fP to send
such log information to stdout.
.TP
\fB\-V\fP
Selects verbose mode, which will send progress reports to stderr [Default runs "silently"].
.TP
\fB\-4\fP
By default, the storage types used in a MGD77+ netCDF file greatly exceed the precision
imposed by the ASCII MGD77 format.  However, for the five items \fBfaa\fP, \fBeot\fP,
\fBmag\fP, \fBdiur\fP and \fBmsd\fP we use 2-byte integers with implied precisions of 0.1 mGal,
0.1 nTesla, and 1 m as in the MGD77 format.  It is possible that at some point these items will
need to be stored as 4-byte ints which would allow precisions of 1 fTesla, 1 nGal, and 0.01 mm,
respectively.  This option activates such storage [Default uses 2-byte integers].
.SH EXAMPLES
To convert 01010047.mgd77 and 01010008.mgd77 to new netCDF .nc files, and
capture all verification messages, try
.br
.sp
\fBmgd77convert\fP 01010047 01010008 \fB\-Fa\fP \fB\-Tc\fP \fB\-V\fP \fB\-Lew+\fP > log.lis
.br
.sp
To convert 01010047.nc back to MGD77 ASCII and make sure it is identical to the
original file, try (Bourne shell syntax)
.br
.sp
orig=`\fBmgd77path\fP 01010047 \fB\-Ic\fP`
.br
\fBmgd77convert\fP 01010047 \fB\-Fc\fP \fB\-Ta\fP \fB\-V\fP
.br
diff $orig 01010047.mgd77
.br
.sp
To convert 01010047.nc to a plain ASCII table for manual editing, overwriting any existing table, try
.br
.sp
\fBmgd77convert\fP 01010047 \fB\-Fc\fP \fB\-T+t\fP \fB\-V\fP
.br
.sp
To recover the original NGDC MGD77 version of 01020051.nc and ignore any E77 corrections,
use
.br
\fBmgd77convert\fP 01020051 \fB\-FC\fP \fB\-Ta\fP \fB\-V\fP
.br
.SH FILE FORMATS
\fBmgd77convert\fP handles four different formats. (1) The MGD77 ASCII tables are
the conventional standard for distribution of underway geophysical data to and from the
NGDC data center.  Normally, only the ship-operations people and the cruise PI might
be involved in \fImaking\fP an MGD77 ASCII file for transmission to NGDC; users
are more interested in \fIreading\fP such files. (2) The MGD77+ netCDF format was
developed to fascilitate the use of MGD77 data by scientists.  It contains all the information
of the original MGD77 file and if you convert back and forth you end up with the original.
However, file sizes are typically ~30% of the original ASCII format and is much faster to
operate on. (3) NGDC has now started to use a new tab-delimited version of the MGD77 data,
called MGD77T.  In addition to all the info in old MGD77 files it contains a few more quality
flags for grav, mag, and bathymetry.  (4) The plain ASCII tab-separated dump is available for users who need to
manually edit the content of a MGD77 file.  This is usually easier to do when the columns
are tab-separated than when they are all crunched together in the MGD77 punch-card format.  It
differs from format MGD77T in that missing items are written as NaNs.
.SH OTHER TOOLS
The MGD77+ netCDF files are CF-1.0 and COARDS compliant and can be examined with general-purpose
tools such as ncBrowse and ncView.
.SH "SEE ALSO"
.IR mgd77manage (1),
.IR mgd77list (1),
.IR mgd77sample (1),
.IR mgd77track (1)
.IR x2sys_init (1)
.SH REFERENCES
Wessel, P., and W. H. F. Smith, 2015, The Generic Mapping Tools (GMT) version
4.5.14 Technical Reference & Cookbook, SOEST/NOAA.
.br
Wessel, P., and W. H. F. Smith, 1998, New, Improved Version of Generic Mapping
Tools Released, EOS Trans., AGU, 79 (47), p. 579.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, EOS Trans., AGU, 76 (33), p. 329.
.br
Wessel, P., and W. H. F. Smith, 1995, New Version of the Generic Mapping Tools
Released, http://www.agu.org/eos_elec/95154e.html, Copyright 1995 by the
American Geophysical Union.
.br
Wessel, P., and W. H. F. Smith, 1991, Free Software Helps Map and Display Data,
EOS Trans., AGU, 72 (41), p. 441.
.br
ncBrowse, see \fIhttp://www.epic.noaa.gov/java/ncBrowse/\fP
.br
ncView, see \fIhttp://meteora.ucsd.edu/~pierce/ncview_home_page.html\fP
.br
The Marine Geophysical Data Exchange Format - "MGD77", see
\fIhttp://www.ngdc.noaa.gov/mgg/dat/geodas/docs/mgd77.txt\fP
.br


gmt-4.5.14/ChangeLog
-----------------------------------------------------------------------------------------
	$Id: ChangeLog 10331 2015-10-29 19:41:56Z pwessel $
-----------------------------------------------------------------------------------------

-----------------------------------------------------------------------------------------
2015-11-01	pwessel
	* GMT 4.5.14 finalized and released
-----------------------------------------------------------------------------------------
2015-10-29	pwessel
	* meca/psmeca.c:	Enforce rakes and strikes to be in -180/+180 range.
				Issue #799, for gmt4 only as not a problem in gmt5.

2015-07-28	pwessel
	* meca/psmeca.c:	Be able to read any value x and y coordinates

2015-06-24	pwessel
	* gmt_support.c:	Fix bug in jump near Dateline in gmt_inonout_sphpol_count,
				affecting inside/outside spherical cases, reported in message-2219

2015-03-18	pwessel
	* share/cpt/GMT_wysiwyg.cpt:	Fix typo in blue for high end [Issue #689].

2015-01-16	pwessel
	* gmt_plot.c:		Fix issue # 662 with graph arrows for negative scales.

2015-01-10	pwessel
	* gmt_plot.c:		Flipping the n/s sides array when lon and -y axes
				fixes issue #520

2015-01-01	pwessel
	* configure.ac:		Moved to 4.5.14 and back to svn release only.
-----------------------------------------------------------------------------------------
2015-01-01	pwessel
	* GMT 4.5.13 finalized and released [GSHHG 2.3.4 was released on Jan 1 as well]
-----------------------------------------------------------------------------------------
2014-12-28	pwessel
	* gmt_map.c:		Needed separate GMT_cartesian_overlap for non-geographic
				projections.  Fixes issue # 657.

2014-09-29	pwessel
	* gmt_plot.c:		Left a PSL variable undefined when grdimage plotted a grid
				with x = longitude and y = Cartesian.

2014-09-09	pwessel
	* grdfilter.c:		Fix y-shifts when -I sets smaller output.  Issue #616.
	* gmt_bcr.c:		Fix problem with points slightly outside causing an
				access violation; see issue #620.

2014-08-29	pwessel
	* sample1d.c:		Fix resampling for decreasing t values.  Issue #608.

2014-08-05	pwessel
	* gmt_map.c:		Could run into trouble clipping polygons for non-periodic
				maps.  Added more checks in GMT_wesn_clip.

2014-07-31	pwessel
	* gmt_map.c:		Rect clipping with -R<w/s/e/n>r could introduce stray lines
				when dataset has features about 180 degrees away from
				the user's area. See issue # 594.

2014-07-30	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.2 (released August 1, 2014)
	* psxy[z].c:		Fixed wrap-around issue in fault lines.

2014-07-27	pwessel
	* gmtselect.c:		The -Z option needed to pass records with NaNs.

2014-07-19	pwessel
	* x2sys/x2sys_solve.c:	Fix problems that arise when a track has less crossovers
				than model parameters.
	* gmtmath.c:		Fix issue #123 with files and -C.

2014-07-01	pwessel
	* configure.ac:		Set default GSHHG version to 2.3.1 (released July 1, 2014)
	
2014-06-28	pwessel
	* mgd77/mgd77sniffer.c:	Needed to check that grid values used in decimate function
	 			also were within reasonable range.

2014-05-26	pwessel
	* gmt_map.c:		The map region clip path for -JE/e was hardwired to a full
				circle and did not care about -R setting.
				When -R...r and -Je/E was given the search for enclosing
				geographic boundaries failed when either the S or N pole
				was the projection's antipole.
2014-05-25	pwessel
	* grdmath.c:		Fix issue #566 bug in grdmath's KM2DEG operator.

2014-03-19	pwessel
	* gmtselect.c:		The -Fpolygon option might miss points because the io-machinery
				would sometimes set the min/max longitudes found incorrectly.

2014-03-14	pwessel
	* gmt_shore.c:		Give more error feedback if nc_open fails due to old library.

2014-03-01	pwessel
	* configure.ac:		Moved to 4.5.13 and back to svn release only.

-----------------------------------------------------------------------------------------
2014-03-01	pwessel
	* GMT 4.5.12 finalized and released [GSHHG 2.3.0 was released on Feb 1]
-----------------------------------------------------------------------------------------
2014-02-17	fwobbe
	* gmt_init.c etc.:	File locking -- if enabled via -DFLOCK -- is
				non-compulsory now.  This means if fcntl fails for some reason,
				GMT continues to access the file.  Hence, it is recommended to
				configure GMT with --enable-flock even if fcntl might be buggy
				on some ancient NFS shares.

2014-02-17	pwessel
	* gmt_map.c:		Size for memory allocation cast the product of nx * ny to
				size_t but overflow had already happen as calculation was
				done using ints, then cast to size_t. Must cast nx and ny
				separately.
2014-02-10	jluis
	* misc/gmt2kml.c:	Make sure the default style IDs are unique for each process.

2014-02-07	jluis
	* ps2raster.c:		Make sure that images in KML files do not carry a directory name

2014-02-03	pwessel
	* pslib.c:		Auto-widen paragraph width to at least fit widest word.

2014-02-01	pwessel
	* configure.ac:		Since GSHHG 2.3.0 was released today, make that the default

2014-01-26	pwessel
	* gmt_grdio.c:		Building the grid command string could exceed its length if
				the argument became exactly the max length of the array; then
				a final extra character \0 would be added and exceed array length.
				Thanks to Joachim Saul for pointing it out.

2014-01-22	pwessel
	* pscoast.c:		The processing of -I -N levels and pens could get out of sync
				so that the wrong pen was used for the specified feature.

2014-01-16	remko
	* pslib.c:		Take superscripts and subscripts into account when determining
				dimensions of text box.

2014-01-10	pwessel
	* grdmath.c:		Had to let SDIST return km instead of degrees since it cannot
				do degrees for geodesics.  By doing km it is now aligned with
				what PDIST and LDIST returns for geographic data, and matches
				what GMT5 does.  Some scripts had to be modified to handle the
				new behavior.  If you use SDIST you may wish to check your usage.
				Also fixed bug in SDIST in that spherical selected geodesics

gmt-4.5.14/src/mgd77/mgd77convert.c
/*--------------------------------------------------------------------
 *	$Id: mgd77convert.c 10289 2014-12-28 21:17:06Z pwessel $
 *
 *    Copyright (c) 2005-2015 by P. Wessel
 *    See README file for copying and redistribution conditions.
 *--------------------------------------------------------------------*/
/*
 * mgd77convert allows for format conversions between three file formats:
 * a) The standard MGD-77 ASCII punchcard file format from NGDC
 * c) An enhanced "MGD77+" format based on netCDF that allows extra columns
 * t) A plain ASCII table version of the MGD-77 punch cards
 *
 * Input files are sought from both current directory and the list of data
 * directories given in the mgd77_paths.txt file in $MGD77_HOME.  Output is
 * always written to the current directory.  No file will be overwritten
 * unless this is requested.
 *
 * Author:	Paul Wessel
 * Date:	10-MAR-2006
 * Version:	1.1
 *
 */
 
#include "mgd77.h"

int main (int argc, char **argv)
{
	int code_pos, i, in_out, argno, n_cruises = 0, n_paths, format[2] = {MGD77_NOT_SET, MGD77_NOT_SET};
		
	GMT_LONG error = FALSE, force = FALSE, high_resolution = FALSE, original = FALSE;
	
	char file[BUFSIZ], **list = NULL, *fcode = "actm";
	char *format_name[MGD77_N_FORMATS] = {"MGD77 ASCII", "MGD77+ netCDF", "ASCII table", "MGD77T ASCII"};

	struct MGD77_CONTROL M;
	struct MGD77_DATASET *D = NULL;
	
	argc = (int)GMT_begin (argc, argv);		/* Initialize GMT Machinery */
	
	/* Initialize MGD77 output order and other parameters*/
	
	MGD77_Init (&M);			/* Initialize MGD77 Machinery */
	
	for (argno = 1; !error && argno < argc; argno++) {	/* Process input options */
		if (argv[argno][0] != '-') continue;
		
		in_out   = 0;	/* 0 for input and 1 for output (see -T -F) */
		code_pos = 2;	/* argv[arg_no][code_pos] holds the character code for format */
		switch (argv[argno][1]) {

			case 'V':
			case '\0':
				error += GMT_parse_common_options (argv[argno], NULL, NULL, NULL, NULL);
				break;
			
			case 'L':	/* Determine level of error/warning checking and log destination */
				M.verbose_level = 0;
				for (i = 2; argv[argno][i]; i++) {
					if (argv[argno][i] == 'e') M.verbose_level |= 2;
					if (argv[argno][i] == 'w') M.verbose_level |= 1;
					if (argv[argno][i] == '+') M.verbose_dest = 1;
				}
				break;
			case 'T':
				in_out = 1;	/* The "to" field is 1; the fall-through via a missing "break" is intentional */
				if (argv[argno][code_pos] == '+') force = TRUE, code_pos++;	/* Force overwriting existing files */
			case 'F':
				switch (argv[argno][code_pos]) {									
					case 'a':		/* Standard ascii MGD77 file */
						format[in_out] = MGD77_FORMAT_M77;
						break;
					case 'C':		/* Enhanced MGD77+ netCDF file */
						original = TRUE;	/* Overlook revisions */
					case 'c':
						format[in_out] = MGD77_FORMAT_CDF;
						break;
					case 'm':
						format[in_out] = MGD77_FORMAT_M7T;
						break;
					case 't':		/* Plain ascii dat table */
						format[in_out] = MGD77_FORMAT_TBL;
						break;
					default:
						fprintf (stderr, "%s: Option -%c Bad format (%c)!\n", GMT_program, argv[argno][1], argv[argno][code_pos]);
						exit (EXIT_FAILURE);
						break;
				}
				break;
	
			case '4':	/* Selected high-resolution 4-byte integer MGD77+ format for mag, diur, faa, eot [2-byte integer] */
				high_resolution = TRUE;
				break;
			default:		/* Options not recognized */
				error = TRUE;
				fprintf (stderr, "%s: Option -%c not recognized!\n", GMT_program, argv[argno][1]);
				break;
		}
	}
	
	if (error) exit (EXIT_FAILURE);

	/* Check that the options selected are mutually consistent */
	
	if (GMT_give_synopsis_and_exit || argc == 1) {	/* Display usage */
		fprintf(stderr,"mgd77convert %s - Convert MGD77 data to other file formats\n\n", MGD77_VERSION);
		fprintf (stderr, "usage: mgd77convert <cruise(s)> -Fa|c|m|t -T[+]a|c|m|t [-L[e][w][+]] [-V] [-4]\n\n");
         
		if (GMT_give_synopsis_and_exit) exit (EXIT_FAILURE);
              
		MGD77_Cruise_Explain ();
		fprintf (stderr, "	[Files are read from data repositories and written to current directory]\n");
		fprintf (stderr, "	-F Convert from a file that is either (a) MGD77 ASCII, (c) MGD77+ netCDF, (m) MGD77T ASCII, or (t) plain table\n");
		fprintf (stderr, "	   Use -FC to recover the original MGD77 setting from the MGD77+ file [Default applies E77 corrections]\n");
		fprintf (stderr, "	-T Convert to a file that is either (a) MGD77 ASCII, (c) MGD77+ netCDF, (m) MGD77T ASCII, or (t) plain table\n");
		fprintf (stderr, "	   By default we will refuse to overwrite existing files.  Prepend + to override this policy.\n");
		fprintf (stderr, "	OPTIONS:\n\n");
		fprintf (stderr, "	-L Log level and destination setting for verification reporting.  Append a combination\n");
		fprintf (stderr, "	   of w for warnings, e for errors, and + to send log to stdout [Default is stderr])\n");
		fprintf (stderr, "	-V verbose, report cruise being processed and error summary\n");
		fprintf (stderr, "	-4 Selects high-resolution, 4-byte storage for mag, diur, faa, eot, and msd with precision\n");
		fprintf (stderr, "	   of 10 fTesla, 1 nGal, 0.01 mm [Default is 2-byte with 0.1 nTesla, 0.1 mGal, m precision]\n");
		exit (EXIT_FAILURE);
	}

	n_paths = MGD77_Path_Expand (&M, argv, argc, &list);	/* Get list of requested IDs */

	if (n_paths == 0) {
		fprintf(stderr, "%s: ERROR: No cruises given\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	if (format[GMT_IN] == MGD77_NOT_SET) {
		fprintf(stderr, "%s: Must specify format of input files\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	
	if (format[GMT_OUT] == MGD77_NOT_SET) {
		fprintf(stderr, "%s: Must specify format of output files\n", GMT_program);
		exit (EXIT_FAILURE);
	}
	
	if (gmtdefs.verbose && format[GMT_IN] == format[GMT_OUT]) {
		fprintf(stderr, "%s: Warning: The two formats chosen are the same\n", GMT_program);
	}
	
	if (format[GMT_OUT] == MGD77_FORMAT_TBL && !(strcmp (gmtdefs.d_format, "%lg") & strcmp (gmtdefs.d_format, "%g"))) {
		strcpy (gmtdefs.d_format, "%.10g");	/* To avoid loosing precision upon rereading this file */
	}
	
	if (format[GMT_OUT] == MGD77_FORMAT_CDF && high_resolution) MGD77_select_high_resolution ();
	
	for (argno = 0; argno < n_paths; argno++) {		/* Process each ID */
	
		D = MGD77_Create_Dataset ();	/* Get data structure w/header */
		MGD77_Reset (&M);		/* Reset to start fresh for next file */

		M.format  = format[GMT_IN];	/* Set input file's format and read everything into memory */
		M.original = original;
		if (original) M.use_corrections[MGD77_M77_SET] = M.use_corrections[MGD77_CDF_SET] = FALSE;	/* Turn off E77 corrections */
		for (i = 0; i < MGD77_N_FORMATS; i++) MGD77_format_allowed[i] = (M.format == i) ? TRUE : FALSE;	/* Only allow the specified input format */
		if (MGD77_Open_File (list[argno], &M, MGD77_READ_MODE)) continue;
		if (MGD77_Read_Header_Record (list[argno], &M, &D->H)) {
			fprintf (stderr, "%s: Error reading header sequence for cruise %s\n", GMT_program, list[argno]);
			exit (EXIT_FAILURE);
		}
		sprintf (file, "%s.%s", M.NGDC_id, MGD77_suffix[format[GMT_OUT]]);
		if (format[GMT_IN] == format[GMT_OUT] && !(M.path[0] == '/' || M.path[1] == ':')) {
			fprintf(stderr, "%s: Input and Output file have same name! Output file will have extension \".new\" appended\n", GMT_program);
			strcat (file, ".new");	/* To avoid overwriting original file */
		}
		if (!access (file, R_OK)) {	/* File exists */
			if (force) {	/* Must delete the file first */
				if (remove (file)) {	/* Oops, removal failed */
					fprintf(stderr, "%s: Unable to remove existing file %s - skipping the conversion\n", GMT_program, file);
					MGD77_Close_File (&M);
					MGD77_Free (D);	/* Free memory allocated by MGD77_Read_File */
					continue;
				}
			}
			else {	/* Cowardly refuse to do this */
				fprintf(stderr, "\n%s: Output file already exists.  Use -T+%c to force overwriting\n", GMT_program, fcode[format[GMT_OUT]]);
				MGD77_Close_File (&M);
				MGD77_Free (D);	/* Free memory allocated by MGD77_Read_File */
				continue;
			}
		}
		
		/* OK, now we can read the data set */
		
		if (MGD77_Read_Data (list[argno], &M, D)) {
			fprintf (stderr, "%s: Error reading data set for cruise %s\n", GMT_program, list[argno]);
			exit (EXIT_FAILURE);
		}
		MGD77_Close_File (&M);
		
		MGD77_Verify_Prep (&M, D);	/* Get key meta-data derived form data records */

		MGD77_Verify_Header (&M, &(D->H), NULL);	/* Verify the header */
	
		if (format[GMT_IN] == MGD77_FORMAT_CDF && format[GMT_OUT] != MGD77_FORMAT_CDF && (D->H.info[MGD77_CDF_SET].n_col || D->flags[0] || D->flags[1])) {
			fprintf(stderr, "\n%s: Warning: Input file contains enhanced material that the output file format cannot represent\n", GMT_program);
		}

		/* OK, ready to write out converted file */
		
		M.format  = format[GMT_OUT];				/* Change the format to the desired output format and write new file in current directory */
		M.original = TRUE;					/* Always write to original attributes */
		for (i = 0; i < MGD77_N_FORMATS; i++) MGD77_format_allowed[i] = (M.format == i) ? TRUE : FALSE;	/* Only allow the specified output format */
		D->H.author = (char *)GMT_memory (VNULL, strlen (M.user)+1, sizeof (char), GMT_program);	/* Allocate space for author */
		strcpy (D->H.author, M.user);									/* Pass current user login id as author */
		if (D->H.history) GMT_free ((void *)D->H.history);	/* Make sure history is blank so it is reset by MGD77_Write_File */
		if (MGD77_Write_File (file, &M, D)) {
			fprintf (stderr, "%s: Error writing new file for cruise %s\n", GMT_program, list[argno]);
			exit (EXIT_FAILURE);
		}
		if (gmtdefs.verbose) {
			fprintf (stderr, "%s: Converted cruise %s to %s format", GMT_program, list[argno], format_name[format[GMT_OUT]]);
			if (D->H.errors[0]) fprintf (stderr, " [%2.2d header problems (%d warnings + %d errors)]", D->H.errors[0], D->H.errors[1], D->H.errors[2]);
			if (D->errors) fprintf (stderr, " [%d data errors]", D->errors);
			fprintf (stderr, "\n");
		}

		MGD77_Free (D);	/* Free memory allocated by MGD77_Read_File */
		n_cruises++;
	}
	
	if (gmtdefs.verbose) fprintf (stderr, "%s: Converted %d MGD77 files\n", GMT_program, n_cruises);
	
	MGD77_Path_Free (n_paths, list);
	MGD77_end (&M);

	exit (EXIT_SUCCESS);
}
